source/Ship.cpp

```diff
@@ -2046,23 +2046,30 @@ void Ship::Fire(vector<Projectile> &projectiles, vector<Visual> &visuals)
 	for(unsigned i = 0; i < hardpoints.size(); ++i)
 	{
 		const Weapon *weapon = hardpoints[i].GetOutfit();
-		if(weapon && CanFire(weapon))
+		if(weapon)
 		{
-			if(weapon->AntiMissile())
-				antiMissileRange = max(antiMissileRange, weapon->Velocity() + weaponRadius);
-			else if(weapon->TractorBeam())
-				tractorBeamRange = max(tractorBeamRange, weapon->Velocity() + weaponRadius);
-			else if(firingCommands.HasFire(i))
+			CanFireResult canFire = CanFire(weapon);
+			if(canFire == CanFireResult::CAN_FIRE)
 			{
-				armament.Fire(i, *this, projectiles, visuals, Random::Real() < jamChance);
-				if(cloak)
+				if(weapon->AntiMissile())
+					antiMissileRange = max(antiMissileRange, weapon->Velocity() + weaponRadius);
+				else if(weapon->TractorBeam())
+					tractorBeamRange = max(tractorBeamRange, weapon->Velocity() + weaponRadius);
+				else if(firingCommands.HasFire(i))
 				{
-					double cloakingFiring = attributes.Get("cloaked firing");
-					// Any negative value means shooting does not decloak.
-					if(cloakingFiring > 0)
-						cloak -= cloakingFiring;
+					armament.Fire(i, *this, projectiles, visuals, Random::Real() < jamChance);
+					if(cloak)
+					{
+						double cloakingFiring = attributes.Get("cloaked firing");
+						// Any negative value means shooting does not decloak.
+						if(cloakingFiring > 0)
+							cloak -= cloakingFiring;
+					}
 				}
 			}
+			else if(ammoAlert && canFire == CanFireResult::NO_AMMO
+					&& firingCommands.HasFire(i) && hardpoints[i].IsReady())
+				Audio::Play(weapon->EmptySound(), position);
```

Comment:
Instead of this, how about:
```c++
		const Weapon *weapon = hardpoints[i].GetOutfit();
		if(!weapon)
			continue;
		CanFireResult canFire = CanFire(weapon);
		if(canFire == CanFireResult::CAN_FIRE)
		{
			if(weapon->AntiMissile())
				antiMissileRange = max(antiMissileRange, weapon->Velocity() + weaponRadius);
			else if(weapon->TractorBeam())
				tractorBeamRange = max(tractorBeamRange, weapon->Velocity() + weaponRadius);
			else if(firingCommands.HasFire(i))
			{
				armament.Fire(i, *this, projectiles, visuals, Random::Real() < jamChance);
				if(cloak)
				{
					double cloakingFiring = attributes.Get("cloaked firing");
					// Any negative value means shooting does not decloak.
					if(cloakingFiring > 0)
						cloak -= cloakingFiring;
				}
			}
		}
		else if(canFire == CanFireResult::NO_AMMO && ammoAlert
				&& firingCommands.HasFire(i) && hardpoints[i].IsReady())
			Audio::Play(weapon->EmptySound(), position);
```